packages/js-dpp/lib/dataContract/stateTransition/DataContractUpdateTransition/validation/basic/validateIndicesAreBackwardCompatible.js

```diff
@@ -0,0 +1,167 @@
+const lodashGet = require('lodash.get');
+const DataContractHaveNewIndexWithOldPropertiesError = require('../../../../../errors/consensus/basic/dataContract/DataContractHaveNewIndexWithOldPropertiesError');
+const DataContractHaveNewUniqueIndexError = require('../../../../../errors/consensus/basic/dataContract/DataContractHaveNewUniqueIndexError');
+const DataContractIndicesChangedError = require('../../../../../errors/consensus/basic/dataContract/DataContractIndicesChangedError');
+
+const serializer = require('../../../../../util/serializer');
+
+const ValidationResult = require('../../../../../validation/ValidationResult');
+
+/**
+ * Check one of old indices have been changed
+ *
+ * @param {string} documentType
+ * @param {object} oldSchema
+ * @param {object[]} newDocuments
+ *
+ * @returns {object}
+ */
+function checkOldIndicesChanged(documentType, oldSchema, newDocuments) {
+  const path = `${documentType}.indices`;
+
+  const newSchemaIndices = lodashGet(newDocuments, path);
+
+  // Building name - index map for easier search
+  const nameIndexMap = (newSchemaIndices || []).reduce((map, indexDefinition) => ({
+    ...map,
+    [indexDefinition.name]: indexDefinition,
+  }), {});
+
+  // Checking every old and it's respective new index
+  // if they are have the same definition
+  const changedIndex = (oldSchema.indices || []).find(
+    (indexDefinition) => (
+      !serializer.encode(indexDefinition).equals(
+        serializer.encode(nameIndexMap[indexDefinition.name]),
+      )
+    ),
+  );
+
+  return changedIndex;
+}
+
+/**
+ * Check if some of the new indices have unique flag
+ *
+ * @param {string} documentType
+ * @param {object} oldSchema
+ * @param {object[]} newDocuments
+ *
+ * @returns {object}
+ */
+function checkNewIndicesAreUnique(documentType, oldSchema, newDocuments) {
+  const newSchemaIndices = lodashGet(newDocuments, `${documentType}.indices`);
+
+  const oldIndexNames = (oldSchema.indices || []).map(
+    (indexDefinition) => indexDefinition.name,
+  );
+
+  // Gather only newly defined indices
+  const newIndices = (newSchemaIndices || []).filter(
+    (indexDefinition) => !oldIndexNames.includes(indexDefinition.name),
+  );
+
+  const indexUnique = (newIndices || []).find((indexDefinition) => indexDefinition.unique === true);
+
+  return indexUnique;
+}
+
+/**
+ * Check if some of the new indices have old properties
+ *
+ * @param {string} documentType
+ * @param {object} oldSchema
+ * @param {object[]} newDocuments
+ *
+ * @returns {object}
+ */
+function checkNewIndicesHaveOldProperties(documentType, oldSchema, newDocuments) {
+  const newSchemaIndices = lodashGet(newDocuments, `${documentType}.indices`);
+
+  const oldIndexNames = (oldSchema.indices || []).map(
```

Comment:
With the new design, you can use old properties at the beginning of an index if they are already indices. The properties sequence must be the same as in existing indices.